PIETOOLS_2021b/opvar2d/opvar2d/@dopvar2d/blkdiag.m

```diff
@@ -0,0 +1,243 @@
+function Fdiag = blkdiag(varargin)
+% F = blkdiag(varargin) generates a block diagonal dopvar2d object using
+% varargs as the diag blocks
+% 
+% INPUTS:
+%   varargs: dopvar2d, opvar2d, dpvar, polynomial, or double object.
+%            At least one object should be opvar2d.
+% 
+% OUTPUTS:
+% Fdiag: a dopvar2d object
+%           - If two objects A,B are dopvar2d/opvar2d, blkdiag(A,B) will perform
+%             concatenation [A.Rij, 0    ]
+%                           [0    , B.Rij] on the different parameters Rij
+%             defining A and B
+%           - If A is opvar2d but B is not, A must have only 1 nonempty
+%             element A.Rij, which will be adjusted to [A.Rij, 0]
+%                                                      [0    , B];
+%           - If B is opvar2d but A is not, B must have only 1 nonempty
+%             element B.Rij, which will be adjusted to [A, 0   ]
+%                                                      [0, B.Rij];
+% 
+% NOTES:
+% For support, contact M. Peet, Arizona State University at mpeet@asu.edu,
+% S. Shivakumar at sshivak8@asu.edu, or Declan Jagt at djagt@asu.edu
+%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Copyright (C)2021  M. Peet, S. Shivakumar, D. Jagt
+%
+% This program is free software; you can redistribute it and/or modify
+% it under the terms of the GNU General Public License as published by
+% the Free Software Foundation; either version 2 of the License, or
+% (at your option) any later version.
+%
+% This program is distributed in the hope that it will be useful,
+% but WITHOUT ANY WARRANTY; without even the implied warranty of
+% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+% GNU General Public License for more details.
+%
+% You should have received a copy of the GNU General Public License
+% along with this program; if not, write to the Free Software
+% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+%
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%
+% If you modify this code, document all changes carefully and include date
+% authorship, and a brief description of modifications
+%
+% Initial coding DJ - 08/27/2021
+
+if nargin==1
+    Fdiag=varargin{1};
+else % sequentially creating a block diagonal matrix
+    A = varargin{1}; B = varargin{2};
+    
+    if isempty(A)
+        Fdiag = B;
+    elseif isempty(B)
+        Fdiag = A;
+    else
+        if (isa(A,'opvar2d') || isa(A,'dopvar2d')) && (isa(B,'opvar2d') || isa(B,'dopvar2d'))
+            if any(any(A.I~=B.I))
+                error('Spatial domain of dopvar2d objects must match for block-diagonal concatenation');
+            elseif any(~isequal(A.var1,B.var1)) || any(~isequal(A.var2,B.var2))
+                error('The spatial variables of dopvar2d objects must match for block-diagonal concatenation');
+            end
+            O1 = dopvar2d([],[A.dim(:,1),B.dim(:,2)],A.I,A.var1,A.var2);

```

Comment:
The blkdiag code seems to have a bug when one of the inputs is not opvar2d/dopvar2d. For example 
```
dopvar2d A;
A = dopvar2d(abs(randi(4,4,2)));
B=blkdiag(2,A);
```
Generates the output
```
>> A
A =

    [0] |   [0] |   [0] |   [0] 
    ---------------------------
    [0] | A.Rxx |   [0] | A.Rx2 
    ---------------------------
    [0] |   [0] | A.Ryy | A.Ry2 
    ---------------------------
    [0] | A.R2x | A.R2y | A.R22 

A.Rxx =

    [0] | [0] | [0] 

A.Rx2 =

    [0] | [0] | [0] 

A.Ryy =

    [0] | [0] | [0] 

A.Ry2 =

    [0] | [0] | [0] 

A.R2x =

    [0] | [0] | [0] 

A.R2y =

    [0] | [0] | [0] 

A.R22 =

    [0] | [0] | [0] 
    --------------- 
    [0] | [0] | [0] 
    --------------- 
    [0] | [0] | [0] 
>> B
B =

    [2,0] |    [] |    [] |    [] 
    [0,0] |       |       |       
    -----------------------------
       [] | B.Rxx |    [] | B.Rx2 
    -----------------------------
       [] |    [] | B.Ryy | B.Ry2 
    -----------------------------
       [] | B.R2x | B.R2y | B.R22 

B.Rxx =

    [] | [] | [] 

B.Rx2 =

    [] | [] | [] 

B.Ryy =

    [] | [] | [] 

B.Ry2 =

    [] | [] | [] 

B.R2x =

    [] | [] | [] 

B.R2y =

    [] | [] | [] 

B.R22 =

    [] | [] | [] 
    ------------ 
    [] | [] | [] 
    ------------ 
    [] | [] | [] 
 ```
 I do not think that is correct (or is that the desired block-diagonal output for this function?)
 I think the errors are due to O1, and O2, in lines 65/66.

Reply:
Thanks for pointing that out! Blkdiag of opvar2d with non-opvar2d object should only be allowed if the opvar2d maps to and from only a single function space, so that we assume the non-opvar2d object to map to and from the same function space. However, instead of checking the number of nonzeros in the dimensions of the opvar2d object, the function checked the number of nonzeros in the domain of the opvar2d object, which of course makes no sense. I fixed the issue in both codes, let me know if you run into any other errors.